src/Cassandra/Mapping/TypeConversion/TypeConverter.cs

```diff
@@ -302,22 +332,91 @@ internal Delegate GetToDbConverter(Type pocoType, Type dbType)
+            {
+                Type dbGenericType = dbType.GetTypeInfo().GetGenericTypeDefinition();
+                Type[] dbTypeGenericArgs = dbType.GetTypeInfo().GetGenericArguments();
+                if (pocoType.GetTypeInfo().IsArray)
+                {
+                    // Its an array, convert each element
+                    return ConvertToIEnumerableToDbTypeMethod
+                        .MakeGenericMethod(pocoType.GetTypeInfo().GetElementType(), dbTypeGenericArgs[0])
+                        .CreateDelegateLocal(this);
+                }
+                if (!pocoType.GetTypeInfo().IsGenericType || dbType.GetTypeInfo().IsAssignableFrom(pocoType))
+                {
+                    return null;
+                }
+                Type[] pocoTypeGenericArgs = pocoType.GetTypeInfo().GetGenericArguments();
+                if (dbGenericType == typeof(IEnumerable<>))
+                {
+                    // Its a list or a set but the child types doesn't match
+                    return ConvertToIEnumerableToDbTypeMethod
+                        .MakeGenericMethod(pocoTypeGenericArgs[0], dbTypeGenericArgs[0]).CreateDelegateLocal(this);
+                }
+                if (dbGenericType == typeof(IDictionary<,>))
+                {
+                    return ConvertToIDictionaryToDbTypeMethod
+                        .MakeGenericMethod(pocoTypeGenericArgs[0], pocoTypeGenericArgs[1], dbTypeGenericArgs[0],
+                            dbTypeGenericArgs[1]).CreateDelegateLocal(this);
+                }
+            }
 
             return null;
         }
 
-        // ReSharper disable UnusedMember.Local 
-        // (these methods are invoked via reflection above)
-        private static Dictionary<TKey, TValue> ConvertToDictionary<TKey, TValue>(IDictionary<TKey, TValue> mapFromDatabase)
+        private Func<TPoco, TDatabase> TryFindToDbConverter<TPoco, TDatabase>()
         {
-            return new Dictionary<TKey, TValue>(mapFromDatabase);
+            var converter = FindToDbConverter<TPoco, TDatabase>();
+            if (converter != null)
+            {
+                return (Func<TPoco, TDatabase>)converter;
+            }
+            
+            TDatabase ChangeType(TPoco a)
+            {
+                try
+                {
+                    return (TDatabase) (object) a;
+                }
+                catch (InvalidCastException ex)
+                {
+                    throw new InvalidCastException(
+                        $"Specified cast is not valid: from {a.GetType()} to {typeof(TDatabase)}", ex);
+                }
+            }
+
+            return ChangeType;
         }
 
-        private Dictionary<TKeyResult, TValueResult> ConvertToDictionaryWithCast<TKeySource, TValueSource, TKeyResult, TValueResult>
-            (IDictionary<TKeySource, TValueSource> mapFromDatabase)
+        private IEnumerable<TResult> ConvertToIEnumerableToDbType<TSource, TResult>(IEnumerable<TSource> items)
```

Comment:
I think the name Convert**From**IEnumerableToDbType would be more clear

Reply:
Nice catch!